pandas/core/computation/engines.py

```diff
@@ -93,18 +93,19 @@ def _evaluate(self):
         -----
         Must be implemented by subclasses.
         """
-        pass
+        # mypy complains if we use @abc.abstractmethod, so we do use
+        #  AbstractMethodError instead
+        from pandas.errors import AbstractMethodError
+
+        raise AbstractMethodError(self)
```

Comment:
mypy can afford additional checking since it has some understanding of abstract classes. so I'm not 100% happy with this, if it sets a precedent for dealing with `Cannot instantiate abstract class 'AbstractEngine' with abstract attribute '_evaluate'` style mypy errors.

in this case,
```
_engines: Dict[str, Type[Union[NumExprEngine, PythonEngine]]] = {
    "numexpr": NumExprEngine,
    "python": PythonEngine,
}
```
also passes mypy without changes to the code and only type annotations added.

It also means we retain the modern Python functionality for dealing with abstract methods instead of homegrown. 

wdyt?

Reply:
> It also means we retain the modern Python functionality for dealing with abstract methods instead of homegrown.

The homegrown doesn't really bother me since we use it in a lot of places.  But annotations is an area where I'm going to defer to you pretty consistently.

> `_engines: Dict[str, Type[Union[NumExprEngine, PythonEngine]]]`

Is this supported in py35?  I rebel at this pattern because I always think "mypy should be able to figure this out" (which is only true if this were some kind of frozen_dict, but still).

Reply:
> Is this supported in py35?

#29212 seems to have gone stale. so will need to use  the py3.5 syntax

> I rebel at this pattern because I always think "mypy should be able to figure this out"

It is required in a lot of places and in this case on my branch I currently have

```
_engines: Dict[str, Type[AbstractEngine]] = {
    "numexpr": NumExprEngine,
    "python": PythonEngine,
}
```
but can't remember why it was needed, but it looks like it will need to be updated to account for the abstract method if not done in this PR.